### DISCUSSION TITLE
Help Browser - Filter the page display

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#4001] [Development]
> The Help Browser is a new feature to bring online help to NeoMutt.
> The code is in the [devel/help](https://github.com/neomutt/neomutt/tree/devel/help) branch.
>
> It adds a new config variable: `$help_doc_dir` which should point to a copy of the [`test-doc`](https://github.com/neomutt/test-doc) repo.
> Help can be activated by pressing <kbd>Alt-H</kbd>

### Part 1 - Filter out YAML

Currently, the raw file is displayed in the pager.
This includes the YAML header, which is no use to the reader.

When we create the temporary file for the pager, don't copy the header block (terminated by an empty line).

### Part 2 - Parse the YAML

The YAML header will contain a lot of information that's useful to NeoMutt.
e.g. Keywords, Author

This will be used to aid the searching and display of the help files.

The data needs to be parsed and stored in `struct Email`s private data.

### Part 3 - Format the Display

Currently, the help files are displayed as plain text.
Some of them contain 'liquid' tags like `{{ page.title }}`

Any `page.*` tags should be looked up using the YAML header data and replaced

Optional extra task -- Implement basic markdown:

- Replace `**text**` with bold text
- Replace `[alt-text](url)` with a number and add a footnote with the url
- Replace `:+1:` with unicode 'thumbs up' (and the others :-)

**Comment by zzandland** [comment]
Hello! 
I'd like to contribute with making this feature - I have few experiences in C++ (mostly data structure and algorithm questions) and Java (board games such as Chess / Minesweeper). Can anyone give me some insight on how and where to start? 

David

**Comment by flatcap** [comment]
Hi David.  Welcome to NeoMutt and thanks for the offer of help.

> how and where to start?

This feature isn't in the master branch yet, so you'll need to build your own NeoMutt.

- NeoMutt source code
  `git clone -b devel/help https://github.com/neomutt/neomutt`
- Help docs (sample markdown)
  `git clone https://github.com/neomutt/test-doc`

To build NeoMutt, you'll need the devel libraries for ncurses and a few other things.

- Configure the build - we don't need much
  `./configure --disable-doc`
  This may tell you to install some more devel packages.
- Build NeoMutt
  `make`

- Add this to you NeoMutt config file:
  `set help_doc_dir = "/home/zzandland/test-doc"`
- Run NeoMutt and open the help:
  `./neomutt`
  <kbd>Alt-H</kbd> and <kbd>Enter</kbd>

You should see a fake mailbox.  The 'emails' are being read from the `test-doc` repo.

Currently, they're Markdown, but displayed as plain text, complete with YAML header.
The first task is to hide that header.

The source for the 'help' mailbox is in [help/help.c](https://github.com/neomutt/neomutt/blob/devel/help/help/help.c)

Have a look around, have a think and importantly ask lots of questions.

You can ask here, or on IRC: #neomutt on irc.freenode.net

Thanks!

**Comment by shadofren** [comment]
Hi!

I can start working on this next monday.

**Comment by louis77** [comment]
To bring this issue up to date I dump my findings so far.

Most of the work for parsing the YAML header was already in place in the latest commit df9e0d6d09ea48bacdf1c15df7e16f971f934798.

But there was a little bug that prevented if from working:

https://github.com/neomutt/neomutt/blob/df9e0d6d09ea48bacdf1c15df7e16f971f934798/help/help.c#L463
which was referenced in the following lines:
https://github.com/neomutt/neomutt/blob/df9e0d6d09ea48bacdf1c15df7e16f971f934798/help/help.c#L488

Fixing that, `help_file_header()` can successfully pass the headers found to the `struct Email` from the only caller, which is `help_doc_from`.

- [ ] Define and implement headers we want to use beside `title`, i.e. `author`
- [ ] The `From` address of a help is currently hardcoded to @flatcap 's email address. Is/Will there be a generic default email which we use? Should an `author` header overwrite this email then?

There is an issue that crashes neomutt in `help_file_header()` if a header line is malformed, like not containing `:` or at least one whitespace:

- [x] Prevent `help_file_header()` from crashing when a malformed header line is found. Simply ignore.

To achieve the goal of Part 1 is actually easy. `help_file_header()` already moves FILE to the point after the header. The offset can be stored in `struct Email -> body -> offset` which is used when copying the content of the file to be displayed by the pager:

- [x] Move FILE further to eliminate trailing empty lines after header
- [x] Store beginning of actual content in Help file in `struct Email -> body -> offset`

For Part 3, I couldn't find any code yet that does that, so this remains to be done.

**Comment by louis77** [comment]
Started to work on Part 3 - Format the Display:

- [x] Parse liquid tags with pattern `{{ page.* }}` and replace with data from YAML header, if available
- [ ] Code still need some refinement (memory, docs) and testing
- [ ] Replace Markdown formatting

**Comment by flatcap** [comment]
I've had a look through my notes for all the possible YAML headers I could think of.

**Yes** to:

- ~`layout: concertina`~
    ~Necessary for Jekyll~

- `title: Short Sentence`
    Necessary

- `description: Longer sentence`
    Necessary

- `keywords: drafts, saved`
    Add search matches for related words that aren't in the text

- `tags: imap, gmail`
    High-level grouping of pages

**Maybe** to:

- `sequence: 123`
    Without a sequence number, the pages will be in alphabetical order.

- `pinned: true`
    Adding a 'pinned' attribute could highlight certain pages.

- `author: John Doe <jd@example.com>`
- `maintainer: John Doe <jd@example.com>`
    Perhaps for credit, perhaps for maintenance?

**No** to:

- `status: wip`
    Seems like a good idea, but just seems to draw attention to unfinished pages.

- `last-updated: 2020-09-30`
- `last-updated-by: John Doe <jd@example.com>`
    Nice idea, but just something else to get out-of-date

---

I think we should change the 'from' to `neomutt-devel@neomutt.org`
In the future, if we keep track of the Author/Maintainer, we can make them the 'to' (and possibly 'cc' the mailing list).

---

Other tasks:

Please can you adopt #1626
Add:
- `$help_index_format` 
- `$help_simple_search`

Also:

- Add new Patterns for searching / limiting
- Add sorting
- Eliminate config

These three are quite large jobs, but fortunately @RomeuG has recently done them for the Address Book.
See:

```
git log --author Romeu -- alias/
```

and [Config Elimination](https://github.com/neomutt/neomutt/issues/2622#issuecomment-701314444)

**Comment by flatcap** [comment]
> **Yes** to:
>  `layout: concertina`
>  Necessary for Jekyll

Update: **No**.
I've discovered a way to default the layout in `_config.yml`
